File: code-examples/scala/indexes/text.scala
import org.mongodb.scala._

import scala.concurrent.Await
import scala.concurrent.duration.Duration
import scala.util.{Failure, Success}

object TextSearchBlog extends App {

  val mongoClient: MongoClient = MongoClient("mongodb+srv://<user>:<password>@<cluster-url>?writeConcern=majority")
  val database: MongoDatabase = mongoClient.getDatabase("testDB")
  val collection: MongoCollection[Document] = database.getCollection("blogPosts")

  import collection.BsonDocumentObservable

  import scala.concurrent.ExecutionContext.Implicits.global

  val indexFuture = collection.createIndex(Document("title" -> "text", "body" -> "text")).toFuture()
  
  indexFuture.onComplete {
    case Success(indexName) => {
      println(s"Index created: $indexName")

      val query = Document("$text" -> Document("$search" -> "life ahead"))
      val projection = Document("_id" -> 0, "title" -> 1)

      val resultFuture = collection.find(query).projection(projection).toFuture()

      resultFuture.onComplete {
        case Success(results) => results.foreach(doc => println(doc))
        case Failure(e) => println("Error: " + e)
      }

      Await.result(resultFuture, Duration.Inf)
      
    }
    case Failure(e) => println("Error: " + e)
  }

  Await.result(indexFuture, Duration.Inf)
  
  mongoClient.close()
}

File: code-examples/java-rs/indexes/text.java
import com.mongodb.reactivestreams.client.*;
import org.bson.Document;

public class Main {
    public static void main(String[] args) {
        MongoClient mongoClient = MongoClients.create("mongodb+srv://<user>:<password>@<cluster-url>?writeConcern=majority");
        MongoDatabase database = mongoClient.getDatabase("testDB");
        MongoCollection<Document> collection = database.getCollection("blogPosts");

        Document indexFields = new Document("title", "text").append("body", "text");
        Document weights = new Document("title", 3).append("body", 10);
        Document indexOptions = new Document("default_language", "english").append("weights", weights);

        collection.createIndex(indexFields, indexOptions).subscribe(new OperationSubscriber<>());

        Bson filter = Filters.text("life ahead");
        Bson projection = Projections.fields(Projections.include("title"), Projections.excludeId());

        collection.find(filter).projection(projection).forEach(doc -> System.out.println(doc.getString("title")),
                Throwable::printStackTrace, () -> mongoClient.close());
    }
    
    static class OperationSubscriber<T> implements Subscriber<T> {
        @Override
        public void onSubscribe(Subscription s) {
            s.request(1);
        }

        @Override
        public void onNext(T t) {
            System.out.println("Operation successful");
        }

        @Override
        public void onError(Throwable t) {
            System.out.println("Operation failed");
            t.printStackTrace();
        }

        @Override
        public void onComplete() {
            System.out.println("Operation completed");
        }
    }
}
